tests/test_helpers/test_async_utils.py

```diff
@@ -313,10 +313,18 @@ async def run(self):
         run.stop()
         run.wait_completed(sync=True)
 
+    @pytest.mark.asyncio
+    async def test_timeout_with_wait_completed_async(self):
+        """Test runnable async methods, waiting asynchronously."""
+
+        class TestRun(Runnable):
+            async def run(self):
+                while True:
+                    await asyncio.sleep(1)
+
         run = TestRun()
         run.start()
-        with pytest.raises(asyncio.TimeoutError):
-            await run.wait_completed(timeout=1)
+        await run.wait_completed(timeout=1)
```

Comment:
from 19bf2639a7a762cb2bd508d174476d354fdcf638:


Python 3.10 changed the behaviour of asyncio.wait_for.

Follow the link to the relevant part of the function implementation
in 3.8 and 3.10, respectively:

- https://github.com/python/cpython/blob/3.8/Lib/asyncio/tasks.py#L493-L503
- https://github.com/python/cpython/blob/3.10/Lib/asyncio/tasks.py#L444-L460

In particular, note that before 3.10, in case the task was not completed,
the TimeoutError error was raised anyway. Instead, in 3.10,
after the task is cancelled, the function tries again to return the result
of the task, and if this attempt raises a 'CancelledError' exception,
only then the TimeoutError is raised.

However, since the class Runnable wraps the task so to suppress the
CancelledError exception (see Runnable._run_wrapper), the attempt
to get the return value always succeeds, and so the TimeoutError is not raised.

This behaviour is substantially different from the previous version.
This commit only updates a relevant test such that expected TimeoutError
to be raised. Probably requires a rethinking of the working of "Runnable.wait_completed".

Reply:
A short code snippet sketching the relationship regarding `Runnable.wait_completed`, `Runnable._run_wrapper` and `Runnable.run`:

```python
import asyncio
from contextlib import suppress


async def run():
    """The actual job, Runnable.run."""
    while True:
        await asyncio.sleep(1)


async def run_wrapper():
    """The Runnable._run_wrapper method."""
    with suppress(asyncio.CancelledError):
        await run()


async def wait_completed():
    """The Runnable.wait_completed, async case."""
    task = asyncio.create_task(run_wrapper())
    await asyncio.wait_for(task, timeout=1)


asyncio.run(wait_completed())

```